File: Astar_Swift/Astar_Swift/Astar.swift
//
//  Astar.swift
//  Astar_Swift
//

import Foundation
import UIKit

public class Astar {

    var mWorldYX :[[Int]];
    private var mArrayOpen :[Node];
    private var mArrayClosed :[Node];
    var mArrayPath :[Node];
    var mDestination :CGPoint?;

    public init() {
        mWorldYX = [[]]
        mArrayOpen = []
        mArrayClosed = []
        mArrayPath = []
    }
    
    public func findPath() {
        mDestination = findDestination()
        if mDestination == nil {
            return
        }
        let nodeStart = getStartNode()
        if nodeStart == nil {
            return
        }
        
        var nodeCurrent :Node?
        mArrayOpen.append(nodeStart!)
        while mArrayOpen.count > 0 {
            nodeCurrent = mArrayOpen.remove(at: 0)
            
            if nodeCurrent!.location.equalTo(mDestination!) {
                break
            }
            
            let arrayNeighbours = getNeighbourNodes(node: nodeCurrent!)
            
            for index in arrayNeighbours.indices {
                let nodeNeighbour = arrayNeighbours[index]
                if mArrayClosed.contains(nodeNeighbour) {
                    continue
                }
                if mArrayOpen.contains(nodeNeighbour) {
                    let tempG = nodeCurrent!.G + ConstantsA.MOVE_COST
                    let tempH = abs((nodeNeighbour.location.x - mDestination!.x) + (nodeNeighbour.location.y - mDestination!.y))
                    let tempF = tempG + Float(tempH)
                    if tempF < nodeNeighbour.F {
                        nodeNeighbour.parent = nodeCurrent
                        nodeNeighbour.F = tempF
                    }
                    continue
                }
                
                nodeNeighbour.parent = nodeCurrent
                nodeNeighbour.setF(destination: mDestination!, moveCost: ConstantsA.MOVE_COST)
                mArrayOpen.append(nodeNeighbour)
            }
            mArrayClosed.append(nodeCurrent!)
        }
        
        while nodeCurrent?.parent != nil {
            mArrayPath.insert(nodeCurrent!, at: 0)
            nodeCurrent = nodeCurrent?.parent
        }
    }

    private func findDestination()->CGPoint? {
        let height = mWorldYX.count
        let width = mWorldYX[0].count
        for x in 0..<width {
            for y in 0..<height {
                let symbol = mWorldYX[y][x]
                if symbol == AstarView.kSquareTarget {
                    let point = CGPoint(x: x, y: y)
                    return point
                }
            }
        }
        return nil
    }

    private func getStartNode()->Node? {
        let width = mWorldYX[0].count
        let height = mWorldYX.count
        for x in 0..<width {
            for y in 0..<height {
                let symbol = mWorldYX[y][x]
                if symbol == AstarView.kSquareStart {
                    let node = Node()
                    node.location = CGPoint(x: x, y: y)
                    return node
                }
            }
        }
        return nil
    }

    private func getNeighbourNodes(node :Node)->[Node] {
        var arrayNodes = Array<Node>()
        
        let x = Int(node.location.x)
        let y = Int(node.location.y)
        
        //Moving in 4 directions - East, West, North, South
        //If we can move diagonally, we should add cases for - NorthEast, SouthEast, NorthWest, SouthWest
        var index = x + 1
        if index < mWorldYX[y].count {
            let symbol = mWorldYX[y][index]
            if symbol == AstarView.kSquareSpace || symbol == AstarView.kSquareTarget {
                let node = Node()
                node.location = CGPoint(x: (index), y: y)
                arrayNodes.append(node)
            }
        }
        
        index = x - 1
        if index >= 0 {
            let symbol = mWorldYX[y][index]
            if symbol == AstarView.kSquareSpace || symbol == AstarView.kSquareTarget {
                let node = Node()
                node.location = CGPoint(x: index, y: y)
                arrayNodes.append(node)
            }
        }
        
        index = y + 1
        if index < mWorldYX.count {
            let symbol = mWorldYX[index][x]
            if symbol == AstarView.kSquareSpace || symbol == AstarView.kSquareTarget {
                let node = Node()
                node.location = CGPoint(x: x, y: index)
                arrayNodes.append(node)
            }
        }
        
        index = y - 1
        if index >= 0 {
            let symbol = mWorldYX[index][x]
            if symbol == AstarView.kSquareSpace || symbol == AstarView.kSquareTarget {
                let node = Node()
                node.location = CGPoint(x: x, y: index)
                arrayNodes.append(node)
            }
        }
        
        return arrayNodes
    }
}

File: Astar/src/com/example/AStar.java
package com.example;

import java.util.ArrayList;

public class AStar {
	
	private String[][] mWorldYX;
	private ArrayList<Node> mArrayOpen;
	private ArrayList<Node> mArrayClosed;
	private ArrayList<Node> mArrayPath;
	private Point mDestination;
	
	public AStar() {
		super();
		mArrayOpen = new ArrayList<Node>();
		mArrayClosed = new ArrayList<Node>();
		mArrayPath = new ArrayList<Node>();
	}

	public void findPath() {
		mDestination = findDestination();
		if(mDestination == null) {throw new Error("Aborting there is no destination location set on the map");}
		Node nodeStart = getStartNode();
		if(nodeStart == null) {throw new Error("Aborting there is no start location set on the map");}
				
		Node nodeCurrent = null;
		mArrayOpen.add(nodeStart);
		while(mArrayOpen.size() > 0) {
			nodeCurrent = mArrayOpen.remove(0);
			
			if(nodeCurrent.getLocation().equals(mDestination)) {
				break;
			}
			
			ArrayList<Node> arrayNeighbours = getNeighbourNodes(nodeCurrent);
			
			int count = arrayNeighbours.size();
			for(int x = 0; x < count; x++) {
				Node nodeNeighbour = arrayNeighbours.get(x);
				if(mArrayClosed.contains(nodeNeighbour)) {
					continue;
				}
				if(mArrayOpen.contains(nodeNeighbour)) {
					float tempG = nodeCurrent.getG() + ConstantsA.MOVE_COST;
					float tempH = Math.abs((nodeNeighbour.getLocation().x - mDestination.x) + (nodeNeighbour.getLocation().y - mDestination.y));
					float tempF = tempG + tempH;
					if(tempF < nodeNeighbour.getF()) {
						nodeNeighbour.setParent(nodeCurrent);
						nodeNeighbour.setF(tempF);
					}
					continue;
				}
				
				nodeNeighbour.setParent(nodeCurrent);
				nodeNeighbour.setF(mDestination, ConstantsA.MOVE_COST);
				mArrayOpen.add(nodeNeighbour);
			}
			mArrayClosed.add(nodeCurrent);
		}
				
		while(nodeCurrent.getParent() != null) {
			mArrayPath.add(0, nodeCurrent);
			nodeCurrent = nodeCurrent.getParent();
		}
	}
	
	public String[][] worldCreate(String[] aArrayWorld) {
		String[][] arrayYX = new String[aArrayWorld.length][];
		for(int y = 0; y < aArrayWorld.length; y++) {	
			String currentItem = aArrayWorld[y];
			int count = currentItem.length();
			arrayYX[y] = new String[count];
			for(int x = 0; x < count; x++) {
				arrayYX[y][x] = currentItem.substring(x, x + 1);
		    }
		}
		
		return arrayYX;
	}
	
	public void worldPrint() {
		int height = mWorldYX.length;
		int width = mWorldYX[0].length;
		String strWorld = "";
		for(int y = 0; y < height; y++) {
			String row = "";
			for(int x = 0; x < width; x++) {
				row = row + mWorldYX[y][x];
			}
			strWorld = strWorld + row + "\n";
		}
		System.out.println(strWorld);
	}
	
	private Node getStartNode() {
		int width = mWorldYX.length;
		int height = mWorldYX[0].length;
		for(int x = 0; x < width; x++) {
			for(int y = 0; y < height; y++) {
				String symbol = mWorldYX[y][x];
				if(symbol.equals(ConstantsA.SYMBOL_START)) {
					Node node = new Node();
					node.setLocation(new Point(x, y));
//					node.setH(Math.abs((x - mDestination.x) + (y - mDestination.y)));
//					node.setF(node.getG() + node.getH());
					return node;
				}
			}
		}
		return null;
	}
	
	private Point findDestination() {
		int height = mWorldYX.length;
		int width = mWorldYX[0].length;
		for(int x = 0; x < width; x++) {
			for(int y = 0; y < height; y++) {
				String symbol = mWorldYX[y][x];
				if(symbol.equals(ConstantsA.SYMBOL_DESTINATION)) {
					Point point = new Point(x, y);
					return point;
				}
			}
		}
		return null;
	}
	
	private ArrayList<Node> getNeighbourNodes(Node aNode) {
		ArrayList<Node> arrayNodes = new ArrayList<Node>();
		
		int x = aNode.getLocation().x;
		int y = aNode.getLocation().y;
		
		//Moving in 4 directions - East, West, North, South
		//If we can move diagonally, we should add cases for - NorthEast, SouthEast, NorthWest, SouthWest
		if((x + 1) < mWorldYX[y].length) {
			String symbol = mWorldYX[y][x + 1];
			if(symbol.equals(ConstantsA.SYMBOL_SPACE) || symbol.equals(ConstantsA.SYMBOL_DESTINATION)) {
				Node node = new Node();
				node.setLocation(new Point(x + 1, y));
				arrayNodes.add(node);
			}
		}
		
		if((x - 1) >= 0) {
			String symbol = mWorldYX[y][x - 1];
			if(symbol.equals(ConstantsA.SYMBOL_SPACE) || symbol.equals(ConstantsA.SYMBOL_DESTINATION)) {
				Node node = new Node();
				node.setLocation(new Point(x - 1, y));
				arrayNodes.add(node);
			}
		}
		
		if((y + 1) < mWorldYX.length) {
			String symbol = mWorldYX[y + 1][x];
			if(symbol.equals(ConstantsA.SYMBOL_SPACE) || symbol.equals(ConstantsA.SYMBOL_DESTINATION)) {
				Node node = new Node();
				node.setLocation(new Point(x, y + 1));
				arrayNodes.add(node);
			}
		}
		
		if((y - 1) >= 0) {
			String symbol = mWorldYX[y - 1][x];
			if(symbol.equals(ConstantsA.SYMBOL_SPACE) || symbol.equals(ConstantsA.SYMBOL_DESTINATION)) {
				Node node = new Node();
				node.setLocation(new Point(x, y - 1));
				arrayNodes.add(node);
			}
		}
		
		return arrayNodes;
	}
	
	public String[][] getWorldYX() {return mWorldYX;}
	public void setWorldYX(String[][] aWorldYX) {mWorldYX = aWorldYX;}
	
	public ArrayList<Node> getPath() {return mArrayPath;}
	
	public Point getDestination() {return mDestination;}
}
